asttokens/asttokens.py

```diff
@@ -209,86 +258,159 @@ def get_tokens(self, node, include_extra=False):
+      # Include leading indentation for multiline statements.
+      start_col_offset = 0
+    else:
+      start_col_offset = self._line_numbers.from_utf8_col(start_node.lineno, start_node.col_offset)
 
-    # To match the behavior of get_text_range, we exclude trailing semicolons and comments.
+    start = (start_node.lineno, start_col_offset)
+
+    # To match the token-using behaviour, we exclude trailing semicolons and comments.
     # This means that for blocks containing multiple statements, we have to use the last one
     # instead of the actual node for end_lineno and end_col_offset.
     end_node = last_stmt(node)
-    end = self._line_numbers.utf8_to_offset(
-      cast(int, end_node.end_lineno),
-      cast(int, end_node.end_col_offset),
-    )
+    end_lineno = cast(int, end_node.end_lineno)
+    end_col_offset = cast(int, end_node.end_col_offset)
+    end_col_offset = self._line_numbers.from_utf8_col(end_lineno, end_col_offset)
+    end = (end_lineno, end_col_offset)
 
     return start, end
 
-  def get_text_unmarked(self, node):
-    # type: (ast.AST) -> str
+  def get_text_positions(self, node, padded):
+    # type: (AstNode, bool) -> Tuple[Tuple[int, int], Tuple[int, int]]
     """
-    Like get_text(), but works without requiring mark_tokens() to have been called.
-    Requires Python 3.8+. Doesn't support astroid.
+    Returns two ``(lineno, col_offset)`` tuples for the start and end of the given node.
+    If the positions can't be determined, or the nodes don't correspond to any particular text,
+    returns ``(1, 0)`` for both.
+
+    ``padded`` corresponds to the ``padded`` argument to ``ast.get_source_segment()``.
+    This means that if ``padded`` is True, the start position will be adjusted to include
+    leading whitespace if ``node`` is a multiline statement.
     """
-    start, end = self.get_text_range_unmarked(node)
-    return self._text[start:end]
+    if getattr(node, "_broken_positions", None):
+      return (1, 0), (1, 0)
+
+    if supports_no_tokens(node):
+      return self._get_text_positions_no_tokens(node, padded)
+
+    return self.asttokens.get_text_positions(node, padded)
+
+
+# Node types that _get_text_positions_no_tokens doesn't support. Only relevant for Python 3.8+.
+_unsupported_no_tokens_types = ()  # type: Tuple[Type[ast.AST], ...]
+if sys.version_info[:2] >= (3, 8):
+  _unsupported_no_tokens_types += (
+    # no lineno
+    ast.arguments, ast.withitem,
+  )
+  if sys.version_info[:2] == (3, 8):
+    _unsupported_no_tokens_types += (
+      # _get_text_positions_no_tokens works incorrectly for these types due to bugs in Python 3.8.
+      ast.arg, ast.Starred,
+      # no lineno in 3.8
+      ast.Slice, ast.ExtSlice, ast.Index, ast.keyword,
+    )
+
+
+def supports_no_tokens(node=None):
+  # type: (Any) -> bool
+  """
+  Returns True if the Python version and the node (if given) are supported by
+  the ``get_text*`` methods of ``ASTText`` without falling back to token
```

Comment:
Is the point performance? Or better behavior? Either way, an explanation of purpose would be helpful.